hoot-core/src/main/cpp/hoot/core/io/OsmApiChangeset.cpp

```diff
@@ -189,6 +188,98 @@ void XmlChangeset::loadElements(QXmlStreamReader& reader, ChangesetType changese
   }
 }
 
+void XmlChangeset::splitLongWays(long maxWayNodes)
+{
+  //  Negative numbers means there is no max set
+  if (maxWayNodes < 0)
+    return;
+  //  Iterate all of the changeset types except delete
+  for (int type = TypeCreate; type < TypeDelete; ++type)
+  {
+    for (XmlElementMap::iterator it = _ways[type].begin(); it != _ways[type].end(); ++it)
+    {
+      XmlWay* way = dynamic_cast<XmlWay*>(it->second.get());
+      while (way->getNodeCount() > maxWayNodes)
+      {
+        //  Create a copy of the way
+        XmlWayPtr newWay(new XmlWay(*way));
+        newWay->changeId(getNextWayId());
+        //  Remove maxWayNodes from the original and add them to this way
+        way->removeNodes(0, maxWayNodes - 1);
+        newWay->removeNodes(maxWayNodes);
+        //  Save the new way
+        _idMap.addId(ElementType::Way, newWay->id());
+        _ways[TypeCreate][newWay->id()] = newWay;
+        _allWays[newWay->id()] = newWay;
+      }
+    }
+  }
+}
+
+void XmlChangeset::fixMalformedInput()
+{
+  //  Element adds cannot have a positive ID, must be negative
+  for (XmlElementMap::iterator it = _relations[TypeCreate].begin(); it != _relations[TypeCreate].end(); ++it)
+  {
+    long old_id = it->first;
+    if (old_id > 0)
+    {
+      //  Get a new negative relation ID
+      long new_id = getNextRelationId();
+      //  Replace old_id everywhere with new_id, in the relation and in other relations
+      replaceRelationId(old_id, new_id);
+    }
+  }
+  for (XmlElementMap::iterator it = _ways[TypeCreate].begin(); it != _ways[TypeCreate].end(); ++it)
+  {
+    long old_id = it->first;
+    if (old_id > 0)
+    {
+      //  Get a new negative way ID
+      long new_id = getNextWayId();
+      //  Replace old_id everywhere with new_id, in the way and in any relations
+      replaceWayId(old_id, new_id);
+    }
+  }
+  for (XmlElementMap::iterator it = _nodes[TypeCreate].begin(); it != _nodes[TypeCreate].end(); ++it)
+  {
+    long old_id = it->first;
+    if (old_id > 0)
+    {
+      //  Get a new negative node ID
+      long new_id = getNextNodeId();
+      //  Replace old_id everywhere with new_id, in the node, and in any ways and relations
+      replaceNodeId(old_id, new_id);
+    }
+  }
```

Comment:
The first half of this function checks for `create` nodes/ways/relations that have positive IDs and replaces them with new, unique negative IDs.

Reply:
I''m curious, If we have a positive ID, doesn't that mean that it should be Modify instead of a create?

Reply:
Maybe, but that would be really bad.  A modification should be under the `<modify>` section and not under the `<create>` section of the `.osc` file.  From here (https://wiki.openstreetmap.org/wiki/OsmChange):
```
<osmChange version="0.6" generator="acme osm editor">
    <create>
        <node id="-1" changeset="43" version="1" lat="-33.9133123" lon="151.1173123" />
        <node id="-2" changeset="43" version="1" lat="-33.9233321" lon="151.1173321" />
        <way id="-3" changeset="43" version="1">
            <nd ref="-1"/>
            <nd ref="-2"/>
        </way>
    </create>
</osmChange>
```
All negative because they are under `<create>` while:
```
<osmChange version="0.6" generator="acme osm editor">
    <modify>
        <node id="1234" changeset="42" version="2" lat="12.1234567" lon="-8.7654321">
            <tag k="amenity" v="school"/>
        </node>
    </modify>
</osmChange>
```
have positive IDs because it is in the `<modify>` section.  Do we want to try to switch any create to modify if the value is positive?  We could easily do that but I think that is probably a big leap to assume that is what is happening.  I think that is is probably just bad IDs rather than being in the wrong section of the file.  I could be wrong though.

Reply:
I don't know.
I was thinking through the flow: Hoot does stuff -> Create changeset -> Apply changeset and wondering how we would get to this situation.
I had not considered the read and apply a random changeset option where we don't know who/what created the changeset so it could have all sorts of funkyness.

Reply:
From hootenanny we should never see this situation.  If we are then we need to fix `changeset-derive` code.  I was thinking of other changesets in this case.